File: leetcode2156/leetcode2156.py
class Solution:
    def subStrHash(self, s: str, power: int, mod: int, k: int, hashValue: int) -> str:
        n = len(s)
        # 用秦九韶算法计算 s[n-k:] 的哈希值
        hash = 0
        for i in range(n - 1, n - k - 1, -1):
            hash = (hash * power + (ord(s[i]) & 31)) % mod
        ans = n - k if hash == hashValue else 0
        pk = pow(power, k, mod)
        # 向左滑窗
        for i in range(n - k - 1, -1, -1):
            # 计算新的哈希值
            hash = (hash * power + (ord(s[i]) & 31) - pk * (ord(s[i + k]) & 31)) % mod
            if hash == hashValue:
                ans = i
        return s[ans: ans + k]

File: leetcode2156/leetcode2156.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
    public:
        string subStrHash(string s, int power, int modulo, int k, int hashValue) {
            int n = s.size();
            long long hash = 0, pk = 1;
            for (int i = n - 1; i >= n - k; i--){
                hash = (hash * power + (s[i] & 31)) % modulo;
                pk = pk * power % modulo;
            }
            int ans = hash == hashValue ? n - k : 0;
            for (int i = n - k - 1; i >= 0; i--){
                hash = (hash * power + (s[i] & 31) - pk * (s[i+ k] & 31) % modulo + modulo) % modulo;
                if (hash == hashValue){
                    ans = i;
                }
            }
            return s.substr(ans, k);
        }
    };